src/curl.c

```diff
@@ -539,7 +526,107 @@ static int swupd_curl_get_file_full(const char *url, char *filename,
 		unlink(filename);
 	}
 
-	return err;
+	return -status;
+}
+
+/*
+ * Determines what strategy to use based on the return code:
+ * - do not retry: if the fault indicates that the failure isn't transient or
+ *                 is unlikely to be successful if repeated (i.e. disk full)
+ * - retry immediately: if the specific fault reported is unusual or rare, it
+ *                      might have been caused by unusual circumstances, the
+ *                      same failure is unlikely to be repeated (e.g. corrupt
+ *                      network package)
+ * - retry after a delay: if the fault is caused by a transient fault (like
+ *                        network connectivity)
+ */
+static enum retry_strategy determine_strategy(int ret_code)
+{
+	switch (ret_code) {
+	case DOWNLOAD_STATUS_FORBIDDEN:
+	case DOWNLOAD_STATUS_NOT_FOUND:
+	case DOWNLOAD_STATUS_WRITE_ERROR:
+		return DONT_RETRY;
+	case DOWNLOAD_STATUS_RANGE_ERROR:
+	case DOWNLOAD_STATUS_PARTIAL_FILE:
+		return RETRY_NOW;
+	case DOWNLOAD_STATUS_ERROR:
+	case DOWNLOAD_STATUS_TIMEOUT:
+		return RETRY_WITH_DELAY;
+	default:
+		return RETRY_NOW;
+	}
+}
+
+static int retry_download(const char *url, char *filename, struct curl_file_data *in_memory_file, bool resume_ok)
+{
+
+	int current_retry = 0;
+	int sleep_time = retry_delay;
+	int strategy;
+	int ret;
+
+	for (;;) {
+
+		/* download file */
+		ret = swupd_curl_get_file_full(url, filename, in_memory_file, resume_ok);
+
+		if (ret == DOWNLOAD_STATUS_COMPLETED) {
+			/* operation successful */
+			break;
+		} else {
+			/* operation failed, determine retry strategy (only if the content
+			 * URL is not local). The value of ret on error will be < 0 */
+			current_retry++;
+			if (content_url_is_local) {
+				strategy = DONT_RETRY;
+			} else {
+				strategy = determine_strategy(-ret);
+			}
+			switch (strategy) {
+			case DONT_RETRY:
+				/* don't retry just return the failure */
+				return ret;
+			case RETRY_NOW:
+				/* if we have reached the retry limit just return the failure,
+				 * if not try again immediately */
+				if (max_retries) {
+					if (current_retry <= max_retries) {
+						fprintf(stderr, "Retry #%d downloading from %s\n", current_retry, url);
+						continue;
+					} else {
+						fprintf(stderr, "Maximum number of retries reached\n");
+					}
+				} else {
+					fprintf(stderr, "Download retries is disabled\n");
+				}
+				return ret;
+			case RETRY_WITH_DELAY:
+				/* if we have reached the retry limit just return the failure,
+				 * if not, wait for the delay and try again */
+				if (max_retries) {
+					if (current_retry <= max_retries) {
+						if (sleep_time) {
+							fprintf(stderr, "Waiting %d seconds before retrying the download\n", sleep_time);
+						}
+						fprintf(stderr, "Retry #%d downloading from %s\n", current_retry, url);
+						sleep(sleep_time);
```

Comment:
We have too much code duplication here. Instead of using a switch we can use an if just to set the sleep time:

    if (strategy == DONT_RETRY) {
       ....
       return;
    }
    
    if (current_retry <= max_retries) {
        fprintf(stderr, "Retry #%d downloading from %s\n", current_retry, url);
        if (strategy == RETRY_WITH_DELAY) {
            sleep(sleep_time);
            ...
        }
        continue;
....

Reply:
yes, but I just thought that using a switch would look cleaner than using ifs, but if you think having less lines of code is better I can change that code to use ifs instead of a switch

Reply:
It's not about saving lines, but in reducing code duplication. printfs for example are hard to maintain if they are duplicated.